### DISCUSSION TITLE
global usings in ImageSharp 3.0

### DISCUSSION THREAD

**Discussion by Marv51** [SixLabors/ImageSharp#2381] [Q&A] [answered]
Hi, I couldn't find any discussion of the `global using` change in ImageSharp 3.0. 

In PR [2229](https://github.com/SixLabors/ImageSharp/pull/2229), along with a lot of code style changes, a "SixLabors.ImageSharp.props" file was added to the nuget package.

If I understand this correctly, this file will add "implicit using (global using)"s for three ImageSharp namespaces to any project that directly consumes this nuget package.

Because this includes types like "SixLabors.ImageSharp.Color", in our project this results in a lot of "CS0104: 'Color' is an ambiguous reference between 'XXX.Color' and 'SixLabors.ImageSharp.Color'".

The only way to opt-out of this seems to either not use the nuget package, or to disable Implicit usings in our project completely?

I can fix my project by fully qualifying my class names that are ambiguous, but I feel this may not be a positive change? And I am unsure if this is an intended change?

![grafik](https://user-images.githubusercontent.com/471338/222676661-2850d142-d705-4aaf-ab89-62a52ab08735.png)

**Comment by tocsoft** [comment, marked as answer]
as a workaround I would think you should theoretically be able to add
```xml
<ItemGroup>
    <Using Remove="SixLabors.ImageSharp" />
    <Using Remove="SixLabors.ImageSharp.PixelFormats" />
    <Using Remove="SixLabors.ImageSharp.Processing" />
</ItemGroup>
```

**Comment by JimBobSquarePants** [comment]
Yes. This is intentional, yes this is a positive change. 

If you’re using the library then more often than not `Color` is an ImageSharp color only.

**Reply by Marv51** [reply]
Absolutely possible, and in examples and tutorials this certainly looks very elegant. 

I do believe System.Windows.Media.Color, Windows.UI.Color and System.Drawing.Color are somewhat pervasive in WPF, WinForms and WinUI applications, but the workaround from @tocsoft works for me.

Thanks for the quick reply!

**Reply by JimBobSquarePants** [reply]
Yep. Great workaround. I'll add it to the docs.

**Reply by ds5678** [reply]
> this is a positive change

I strongly disagree. If users want an implicit using, they can add it themselves. Putting your library in scope everywhere is ridiculous. No library (not even the BCL) should do that. It is an anti-pattern and should be reverted.

**Reply by JimBobSquarePants** [reply]
>I strongly disagree. If users want an implicit using, they can add it themselves. Putting your library in scope everywhere is ridiculous. No library (not even the BCL) should do that. It is an anti-pattern and should be reverted.

If that was the case, then the functionality and guidance to implement via NuGet package would not exist. Your limited use scope is not indicative of our user base.

This is a warning. DO NOT ever describe our well researched decision making as "ridiculous". You will be banned from all our repositories if you insult our work again.

**Reply by huoyaoyuan** [reply]
I also faced this issue. I'm getting conflict on `ImageInfo` on a project transitively imports ImageSharp.

> If that was the case, then the functionality and guidance to implement via NuGet package would not exist.

Can you refer to the guidance please? Adding implicit usings on transitive dependencies sounds really dangerous to me.

**Reply by ithline** [reply]
I don't see how this is a positive change. After upgrading package version, my project suddenly started throwing errors all around for ambiguous type references. Spent ton of time looking through all files/configs where I could have added global import, but there was nothing. 

Why is it enabled by default? I have never ever seen a nuget package enable this without explicit user confirmation (setting a property). This "feature" puts ImageSharp in context everywhere, breaking existing code that does not depend on ImageSharp just because I happen to have a use case in one small part of the project.
